integreat_cms/cms/utils/slug_utils.py

```diff
@@ -116,69 +120,139 @@ def generate_unique_slug(**kwargs: Unpack[SlugKwargs]) -> str:
+                object_instance,
+                foreign_model,
+                foreign_object,
+                content_models,
             )
+
+        if not other_objects.exists() and not is_reserved_slug(
+            unique_slug, foreign_model
         ):
             break
-        i += 1
-        unique_slug = f"{slug}-{i}"
+
+        counter += 1
+        unique_slug = f"{base_slug}-{counter}"
 
     logger.debug("unique slug: %r", unique_slug)
     return unique_slug
+
+
+def handle_logging(
+    content_models: list[str],
+    foreign_model: str | None,
+    region: Region | None,
+    language: Language | None,
+    slug: str,
+) -> None:
+    """
+    This method handles logging for the slug generation process.
+    """
+    logger.debug("foreign_model: %r", foreign_model)
+    if foreign_model in content_models:
+        logger.debug("%r, %r", region, language)
+    logger.debug("slug: %r", slug)
+
+
+def generate_base_slug(
+    slug: str,
+    fallback: str,
+    cleaned_data: dict[str, Any],
+    object_instance: AbstractBaseModel,
+    foreign_model: str | None,
+) -> str:
+    """
+    Generates the base slug either from the given slug or from the fallback field.
+    """
+    if slug:
+        return slug
+
+    if fallback not in cleaned_data:
+        raise ValidationError(
+            _("Cannot generate slug from {}.").format(_(fallback)),
+            code="invalid",
+        )
+
+    allow_unicode = object_instance._meta.get_field("slug").allow_unicode
+    slug = slugify(cleaned_data[fallback], allow_unicode=allow_unicode)
+
+    return slug or foreign_model or ""
+
+
+def get_prefiltered_queryset(
+    content_models: list[str],
+    manager: Any,
+    foreign_model: str | None,
+    region: Region | None,
+    language: Language | None,
+) -> Any:
+    """
+    Returns a prefiltered queryset depending on the foreign model, region and language.
+    """
+    if foreign_model in content_models:
+        return manager.filter(
+            **{
+                foreign_model + "__region": region,
+                "language": language,
+            }
+        )
+    return manager
+
+
+def is_reserved_slug(slug: str | None, foreign_model: str | None) -> bool:
+    """
+    Checks whether the given slug is reserved for the given foreign model.
+    """
+    if foreign_model == "page":
+        return slug in settings.RESERVED_REGION_PAGE_PATTERNS
+    if foreign_model == "region":
+        return slug in settings.RESERVED_REGION_SLUGS
+    return False
+
+
+def exclude_current_object(
+    qs: QuerySet[T],
+    object_instance: AbstractBaseModel,
+    foreign_model: str | None,
+    foreign_object: AbstractContentModel | None,
+    content_models: list[str],
+) -> QuerySet[T]:
+    """
+    Excludes the current object from the given queryset to avoid false positives when checking for slug uniqueness
+    """
+    if not object_instance:
+        return qs
+
+    if foreign_model in content_models:
```

Comment:
Aah, I remember why that had to be in here, because we cannot just pass objects without an ID with the new django version anymore.
In that case, the foreign object (page, event or location) does not exist in the db yet, meaning that we are trying to create it. When it doesn't exist yet, we also don't have any other slugs in the database to exclude when looking for naming conflicts, wo skipping that step and running the lines following is valid.
```suggestion
    if (
        foreign_model in content_models
        and (foreign_object or object_instance.foreign_object).id
    ):
```